### PR TITLE
Support SAML-auth and mTLS auth for Jolokia

### PR DESCRIPTION
**Pull request by kahowell**
In order to activate jolokia on an actuator, I added a dep for jolokia-core, and modified the application.yaml to enable the endpoint.

I added logging of the principal to the JMX endpoints, which will give us a nice logged record of what principal hit the endpoint. Given that the messages use the toString, the logging will work with either a RhAssociate (internal SAML) or X509 Principal (mTLS) is used.

When either of the two new principal types is used, RBAC is skipped, and a role - `RH_INTERNAL` is added instead.

I changed `IdentityHeaderAuthenticationDetailsSource` to `IdentityHeaderAuthenticationDetailsService`, a change that was
hinted at by @awood. This allowed some separation of creation of the principal and population of the roles. I did not pursue moving the opt-in check into this service in this PR (that looked messy).

I moved to Jackson serialization of the x-rh-identity header object (too much raw JSON manipulation is cumbersome in Java).

To test, launch the application, and then mock some authenticated requests and unauthenticated requests to the jolokia endpoint:

```
DATA='{"type":"exec","mbean":"org.candlepin.subscriptions.jmx:name=tallyJmxBean,type=TallyJmxBean","operation":"tallyAccount(java.lang.String)","arguments":["123456"]}'
RH_ASSOCIATE="x-rh-identity: $(echo '{"identity":{"type":"Associate","associate":{"email":"test@example.com"}}}' | base64 -w0)"
INSIGHTS_USER="x-rh-identity: $(echo '{"identity":{"account_number":"myaccount", "internal":{"org_id":"foobar"}}}' | base64 -w0)"
X509_AUTH="x-rh-identity: $(echo '{"identity":{"type":"X509", "x509":{"subject_dn":"/CN=test.example.com"}}}' | base64 -w0)"
curl -H 'Content-Type: application/json' -d $DATA http://localhost:8080/actuator/jolokia  # anonymous
curl -H 'Content-Type: application/json' -d $DATA -H "$INSIGHTS_USER" http://localhost:8080/actuator/jolokia  # insights user
curl -H 'Content-Type: application/json' -d $DATA -H "$RH_ASSOCIATE" http://localhost:8080/actuator/jolokia  # rh associate
curl -H 'Content-Type: application/json' -d $DATA -H "$X509_AUTH" http://localhost:8080/actuator/jolokia  # mTLS auth
```

Related: RedHatInsights/turnpike#11

### CODE REVIEW

**Comment by kahowell** [comment]
I'd actually recommend testing with turnpike (https://github.com/RedHatInsights/turnpike)

1. Clone turnpike, follow its README for developer setup.
2. Build a fresh rhsm-subscriptions container image: `sudo podman build -t rhsm-subscriptions .`
3. Add to `dev-backends.yml`:
```yaml
- name: rhsm-subscriptions
  route: /rhsm-subscriptions/actuator
  origin: http://rhsm-subscriptions:8081/actuator
  auth:
    saml: "True"
```

4. Add to `docker-compose.yml`:
```yaml
  rhsm-subscriptions:
    image: rhsm-subscriptions
    hostname: rhsm-subscriptions
    ports:
      - 8081:8081
    networks:
      mynet:
        aliases:
          - rhsm-subscriptions
    environment:
      RHSM_SUBSCRIPTIONS_DATASOURCE_PLATFORM: hsqldb
      RHSM_SUBSCRIPTIONS_DATASOURCE_DRIVER_CLASS_NAME: org.hsqldb.jdbc.JDBCDriver
      RHSM_SUBSCRIPTIONS_DATASOURCE_URL: jdbc:hsqldb:mem:testdb
      RHSM_SUBSCRIPTIONS_INVENTORY_SERVICE_DATASOURCE_PLATFORM: hsqldb
      RHSM_SUBSCRIPTIONS_INVENTORY_SERVICE_DATASOURCE_DRIVER_CLASS_NAME: org.hsqldb.jdbc.JDBCDriver
      RHSM_SUBSCRIPTIONS_INVENTORY_SERVICE_DATASOURCE_URL: jdbc:hsqldb:mem:testdb2
      HAWTIO_BASE_PATH: /rhsm-subscriptions/actuator/hawtio
      JAVA_MAX_MEM_RATIO: 85
      GC_MAX_METASPACE_SIZE: 256
      RBAC_USE_STUB: true
      SERVER_PORT: 8081
      AB_JOLOKIA_OFF: true
```
5. Invoke `sudo podman-compose up`
6. Wait a bit for startup (1 minute should be sufficient)
7. Visit https://turnpike.example.com/rhsm-subscriptions/actuator/hawtio

**Review by mstead — CHANGES_REQUESTED** [review]
Looks good and tested well. One question and suggestion.

File: src/main/resources/rhsm-subscriptions.properties [hunk 1]
```diff
@@ -75,3 +75,10 @@ rhsm-subscriptions.rbacApplicationName=${RBAC_APPLICATION_NAME:subscriptions}
 rhsm-subscriptions.rbac-service.useStub=${RBAC_USE_STUB:false}
 rhsm-subscriptions.rbac-service.url=http://${RBAC_HOST:localhost}:${RBAC_PORT:8819}/api/rbac/v1
 rhsm-subscriptions.rbac-service.maxConnections=${RBAC_MAX_CONNECTIONS:100}
+
+###################################
+# Hawtio Console Configuration
+###################################
+hawtio.authenticationEnabled=false
+# disable the remote connection tab, we do not need it
+hawtio.disableProxy=true
```
[the remark this answers is not in the document]
Comment (kahowell) [reply src/main/resources/rhsm-subscriptions.properties]: @mstead, dunno why but can't override w/ env vars by default. Added env vars for these and a few other vars we might tweak locally too.

File: src/main/java/org/candlepin/subscriptions/jmx/OptInJmxBean.java [hunk 2]
```diff
@@ -58,6 +59,8 @@ public String getOptInConfig(String accountNumber, String orgId) {
     @ManagedOperationParameter(name = "accountNumber", description = "Red Hat Account Number")
     @ManagedOperationParameter(name = "orgId", description = "Red Hat Org ID")
     public void optOut(String accountNumber, String orgId) {
+        Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+        log.info("Opt out for {} triggered via JMX by {}", accountNumber, principal);
```
[the remark this answers is not in the document]
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/jmx/OptInJmxBean.java:63]: Sure. Will push a commit with this shortly.

**Review by mstead — APPROVED** [review]

**Comment by awood** [comment]
> I'd actually recommend testing with turnpike (https://github.com/RedHatInsights/turnpike)

Is this something we'd want to add to our standard setup?

**Pull request merged**